proposals/0545-deterministic-operator-parsing.md

```diff
@@ -0,0 +1,150 @@
+---
+author: Las Safin
+date-accepted: ""
+ticket-url: ""
+implemented: ""
+---
+
+This proposal is [discussed at this pull request](https://github.com/ghc-proposals/ghc-proposals/pull/545>).
+
+# Deterministic operator parsing
```

Comment:
Parsing operators is already deterministic. You are probably looking for a different word.

Reply:
Yes, I am, I was not sure what to call it. What do you think is better?

Reply:
It is deterministic in the sense that it doesn't depend on the environment!

Reply:
Lexical operator fixity, perhaps, as you determine their precedence and associativity from their lexical form instead of consulting the corresponding fixity declaration.

Reply:
That does seem like a better name. Unless someone finds a better name this would probably be the name of the extension.

Reply:
> It is deterministic in the sense that it doesn't depend on the environment!

This depends on whether you consider imported modules part of the environment. I see them as part of the program, not the environment. That is, a Haskell program is a _set_ of modules, not a single module, and you need to consider them all together to actually reason about code. Otherwise, you couldn't even tell what type a imported identifier has.

```
import M (x)
```

Would you say that it's non-deterministic whether `x :: Int` or `x :: Bool`? I would surely hope that is deterministic in _my_ builds 😅

Reply:
I would say something like `Pragma-free` instead of `Deterministic`

Reply:
We could call it "pure". A pure function evaluates to the same result in all contexts and program histories. With pure operator fixity, an operator has the same fixity in all contexts and program histories. Just like with functions, purity allows local reasoning.

Reply:
But "purity" is typically taken to be a stronger concept than "determinacy", isn't it? That is nothing can be pure unless it is at least deterministic. If that's so then any argument that rejects "deterministic" as a description of this parsing method ought also to reject "pure".